ngraph/core/src/op/util/arithmetic_reductions_keep_dims.cpp

```diff
@@ -32,61 +32,30 @@ bool ngraph::op::util::ArithmeticReductionKeepDims::visit_attributes(AttributeVi
 
-        if (input_rank.is_static())
-            result_shape = PartialShape::dynamic(input_rank);
-
-        const auto& axes = get_constant_from_source(input_value(1));
-        if (input_rank.is_static() && axes)
-        {
-            AxisSet reduction_axes;
-            auto reduction_axes_val = axes->cast_vector<int64_t>();
-            for (auto axis : reduction_axes_val)
-            {
-                try
-                {
-                    axis = normalize_axis(this, axis, input_rank);
-                }
-                catch (const ngraph_error&)
-                {
-                    NODE_VALIDATION_CHECK(this,
-                                          false,
-                                          "Reduction axis (",
-                                          axis,
-                                          ") is out of bounds ",
-                                          "(argument shape: ",
-                                          input_shape,
-                                          ", reduction axes: ",
-                                          reduction_axes,
-                                          ")");
-                }
-                reduction_axes.insert(axis);
-            }
-
-            std::vector<Dimension> dims;
-            for (int64_t i = 0; i < input_rank.get_length(); i++)
-            {
-                if (reduction_axes.count(i) == 0)
-                {
-                    dims.push_back(input_shape[i]);
-                }
-                else
-                {
-                    dims.emplace_back(Dimension{1});
-                }
-            }
-            result_shape = PartialShape(dims);
-        }
-        set_input_is_relevant_to_shape(1);
-        set_output_type(0, get_input_element_type(0), result_shape);
-    }
-    else
-    {
-        ArithmeticReduction::validate_and_infer_types();
-    }
+    const PartialShape& data_shape = get_input_partial_shape(0);
+    const element::Type& data_et = get_input_element_type(0);
+    const PartialShape& axes_shape = get_input_partial_shape(1);
+    const element::Type& axes_et = get_input_element_type(1);
+
+    NODE_VALIDATION_CHECK(this,
+                          data_et.is_real() || data_et.is_integral_number(),
+                          "Element type of data input must be numeric. Got: ",
+                          data_et);
+
+    NODE_VALIDATION_CHECK(this,
+                          axes_et.is_integral_number(),
+                          "Element type of axes input must be integer. Got: ",
+                          axes_et);
+
+    const Rank axes_rank = axes_shape.rank();
+    NODE_VALIDATION_CHECK(this,
+                          axes_rank.compatible(0) || axes_rank.compatible(1),
+                          "Axes input must be a scalar or 1D input. Got: ",
+                          axes_shape);
+
+    PartialShape result_shape =
+        infer_reduction_output_shape(this, data_shape, axes_shape, m_keep_dims);
```

Comment:
If we need to pass `this` as a first parameter, can we introduce `infer_reduction_output_shape` as a protected method in the base class?

Reply:
Done. Please take a look.